### ISSUE TITLE
Add option for AND operator for the terms in the project search

### ISSUE THREAD

**Issue by bgirardot-np** [hotosm/tasking-manager#675]
It looks like right now the search terms put in the full text search are 'or' so "American Samoa" returns anything with american or samoa.

The search terms should be 'and' so anything with american and samoa.

**Comment by hunt3ri** [comment]
I'd be careful before implementing this, as it means ALL  terms in the description need to be present which in local testing was extremely frustrating for users, as they often got no results from the text search.  Possibly quoting the string would treat it as a whole rather than two separate words?

**Comment by bgirardot** [comment]
@hunt3ri I think the quoted string sounds like a good compromise.

Any chance we could support the word "and" as an operator? e.g., "missing maps" and laos ?

**Comment by bgirardot** [comment]
This issue can be partially addressed in server/services/project_search_service.py at line 132

**Comment by fiennyangeln** [comment]
Hi @bgirardot may I take this issue? i will try looking at the code first

**Comment by fiennyangeln** [comment]
@bgirardot I think the text search need some database trigger to work. Thats why it doesnt work on local dev, including mine. However, when I tried searching for the database trigger in the tm database, there is a lot of trigger in it that I could not differentiate.

**Comment by bgirardot** [comment]
@fiennyangeln Does this information help at all? Especially toward the bottom where it lists the trigger and the procedure that needs to be executed.

tm3=> \d+ project_info
                                    Table "public.project_info"
        Column         |          Type          | Modifiers | Storage  | Stats target | Description 
-----------------------+------------------------+-----------+----------+--------------+-------------
 project_id            | integer                | not null  | plain    |              | 
 locale                | character varying(10)  | not null  | extended |              | 
 name                  | character varying(512) |           | extended |              | 
 short_description     | character varying      |           | extended |              | 
 description           | character varying      |           | extended |              | 
 instructions          | character varying      |           | extended |              | 
 project_id_str        | character varying      |           | extended |              | 
 text_searchable       | tsvector               |           | extended |              | 
 per_task_instructions | character varying      |           | extended |              | 
Indexes:
    "project_info_pkey" PRIMARY KEY, btree (project_id, locale)
    "idx_project_info composite" btree (locale, project_id)
    "textsearch_idx" gin (text_searchable)
Foreign-key constraints:
    "project_info_project_id_fkey" FOREIGN KEY (project_id) REFERENCES projects(id)
Triggers:
    tsvectorupdate BEFORE INSERT OR UPDATE ON project_info FOR EACH ROW EXECUTE PROCEDURE tsvector_update_trigger('text_searchable', 'pg_catalog.english', 'project_id_str', 'short_description', 'description')

Hi @hunt3ri , do you have any ideas or suggestions for what we need to do to get the full text search working in our dev installs?

**Comment by xamanu** [comment]
I think there is no clear assumption whether this should be AND or OR. I would like to support both, and would leave the search as it is with OR, but support AND with a clear syntax:

> american+Samoa

**Comment by ramyaragupathy** [comment]
@prabinoid - can you please confirm if this is still relevant?

cc @suzit-10

**Comment by prabinoid** [comment]
The search text that matches the project the most comes up first and the OR search appears later in the results. I think this is the expected workflow.

**Comment by ramyaragupathy** [comment]
We need to understand:
* in which fields the search team apply - desc, short desc, title, author, imagery source, etc.,
* how is the search executed:
    * the term `mozambique` returns a certains number of proejcts
    * the term `mozambique f` retruns only 1 project, whereas
    * the term `mozambique floods` seems to return abotu 45 projects

So getting clarity on how this is executed internally would help us determine next actions.

cc @prabinoid @suzit-10 @RAytoun

**Comment by ramyaragupathy** [comment]
Sharing the link to @prabinoid's summary on existing search function: https://docs.google.com/document/d/1UC7QplKgXGkBvTNzedfYjfrSREbqVjkZE9rqLEUr3_U/edit?tab=t.0

@prabinoid  -  for records, I am capturing the understanding here, please confirm if this is right.
When a user searches: "Green Energy"


The system:

1. Cleans the text to remove special characters so PostgreSQL query syntax doesn’t break.

2. Builds a full-text search query like:

   ```
   Green & Energy
   ```

   In PostgreSQL, this uses **Full-Text Search**, where:

   * words are reduced to their roots (stemming)
   * so `"energy"` will also match `"energies"`, `"energetic"`, etc.
   * both words must appear somewhere in the indexed text (name or description)

3. Builds a fuzzy name query (substring match) like:

   ```
   %Green Energy%
   ```

   This is used with `ILIKE` and matches the exact character sequence inside the project name.

4. Returns projects where either condition is true:

   * the description (or searchable text) contains both stemmed words
     **OR**
   * the project name contains the exact typed phrase


All of this happens through:
* tsvector indexing
* stemming + tokenization in the full text search
* substring match using ilike
* finally parameter binding to treat input as data rather than executable query

**Comment by ramyaragupathy** [comment]
Input = "green energy"

Project Name | Description Contains | Full-Text Search (green & energy) | Name ILIKE (%green energy%) | Returned?
-- | -- | -- | -- | --
Green Project | only green | ❌ No (missing energy) | ❌ No | ❌ No
Green Project | green + energy | ✅ Yes | ❌ No | ✅ Yes
Energy Project | only energy | ❌ No (missing green) | ❌ No | ❌ No
Green Energy Program | anything | ✅ Yes | ✅ Yes | ✅ Yes
Green Energies Program | anything | ✅ Yes (stemming matches) | ❌ No (phrase differs) | ✅ Yes
Energy Green Program | anything | ✅ Yes (order doesn’t matter) | ❌ No | ✅ Yes
Generator Initiative | unrelated | ❌ No | ❌ No | ❌ No
GreenPower Energy | contains both words separated | ✅ Yes | ❌ No | ✅ Yes

Please see if the above understanding is right.

**Comment by prabinoid** [comment]
Yes, @ramyaragupathy. In short, a DB trigger maintains a tsvector column (text_searchable) built from project_id, project_name, and project_description; the search endpoint sanitizes user input by removing PostgreSQL special operators,  then runs a dual-search: a Full-Text Search where the cleaned query words are joined with & into a tsquery (so all words must match stems/roots across the indexed text), and a fuzzy ILIKE on project_name using %{term}% for simple partial matches; results are returned if either method matches.
